Anchor

garrettjoecox opened the pull request:
![anchor](https://github.com/garrettjoecox/OOT/assets/7316699/a8feac51-47b6-4e4c-b940-2f49fc0bc764)

Please read the known issues & TODO lists below before reporting any bugs or requesting any features.

This is a beta version, be prepared to encounter issues. The stable version can be found [here](https://github.com/garrettjoecox/OOT/pull/64). I am pushing this up now to encourage people to move off of the holiday build, as well as get some broader playtesting on the nightlies in preparation for the 9.0.0 release.

## What is this?

Anchor is a co-op mode for Ship of Harkinian. It allows multiple people to play the game together, vanilla or randomizer, sharing inventory items and world flags.

## Downloads
<!--- section:artifacts:start -->
### Build Artifacts
  - [soh-linux.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/4564569005.zip)
  - [soh-mac.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/4564595768.zip)
  - [soh-windows.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/4564677326.zip)
<!--- section:artifacts:end -->

## How To
- All players should start at the file select screen
- Open network tab, change scheme to "Anchor"
- Agree upon a room ID across all players and click "Enable" (Use something unique, this is basically your password)
- One person should configure the randomizer settings and generate a seed, then share the newly generated JSON spoiler file with other players
- All players should load the same JSON spoiler file (drag it into SoH window), make sure seed icons match, then create a new file.
- If using an existing save/seed ensure the player with the most progress loads the file first.
- All other players should then load their files, and their save state will be synced with the main player's
- After everyone has loaded in, verify on the network tab that it doesn’t warn about anyone being on a wrong version or seed

## TODO/Known Issues
- If you encounter anything not listed here please report it in the [Anchor thread](https://discord.com/channels/808039310850130000/1272968989970141284) in discord
- Beans usage/ammo not synced
- Getting a milk refill will give all other players a bottle with milk
- Giant's Knife get sent across as BGS
- ~~Issues saving/loading presets with the network block~~ Upstream issue, will fix later

aMannus commented on soh/soh/Network/Anchor/Anchor.cpp:
```diff
@@ -0,0 +1,405 @@
+#ifdef ENABLE_REMOTE_CONTROL
+
+#include "Anchor.h"
+#include <nlohmann/json.hpp>
```
I've thrown up this ball on discord recently, but I'd like to put it out there for more public consideration. So far, this has always been called Anchor, which makes sense when you're trying to communicate it being separate from SoH itself. However, now that we're looking to upstream this, I'd personally really like if we rename everything currently called anchor to something more directly descriptive.

For example, my personal recommendation: `Online Multiplayer`.

I know a lot of people know the name Anchor by now, but I think there's going to be issues with people that download a normal SoH build and seeing "Anchor" in the network tab without having the description from the Anchor PR that people have now. Compare that to them seeing "Online Multiplayer", I don't think I have to explain people are going to understand what that means more quickly.

Just food for thought, I don't think I'll ever be in a position to force a change like this especially with Anchor's history, but I do think this is the right move.

aMannus commented on soh/src/code/z_en_item00.c:
```diff
@@ -767,6 +767,13 @@ void EnItem00_Update(Actor* thisx, PlayState* play) {
     EnItem00* this = (EnItem00*)thisx;
     s32 pad;
 
+    // #region SOH [Co-op]
+    if (Flags_GetCollectible(play, this->collectibleFlag)) {
+        Actor_Kill(&this->actor);
+        return;
+    }
+    // #endregion
```
Hook material

aMannus commented on soh/src/overlays/actors/ovl_Bg_Bombwall/z_bg_bombwall.c:
```diff
@@ -211,7 +211,9 @@ void func_8086ED50(BgBombwall* this, PlayState* play) {
 
 void func_8086ED70(BgBombwall* this, PlayState* play) {
-    if (this->collider.base.acFlags & AC_HIT) {
+    // #region SOH [Co-op]
+    if ((this->collider.base.acFlags & AC_HIT) || Flags_GetSwitch(play, this->dyna.actor.params & 0x3F)) {
+    // #endregion
```
Hook material

aMannus commented on soh/src/overlays/actors/ovl_Bg_Haka_Zou/z_bg_haka_zou.c:
```diff
@@ -268,7 +268,9 @@ void func_80882E54(BgHakaZou* this, PlayState* play) {
 void func_80883000(BgHakaZou* this, PlayState* play) {
-    if (this->collider.base.acFlags & AC_HIT) {
+    // #region SOH [Co-op]
+    if ((this->collider.base.acFlags & AC_HIT) || Flags_GetSwitch(play, this->switchFlag)) {
+    // #endregion
         Flags_SetSwitch(play, this->switchFlag);
```
Hook material

aMannus commented on soh/src/overlays/actors/ovl_Bg_Hidan_Dalm/z_bg_hidan_dalm.c:
```diff
@@ -126,8 +126,10 @@ void BgHidanDalm_Destroy(Actor* thisx, PlayState* play) {
 
-    if ((this->collider.base.acFlags & AC_HIT) && !Player_InCsMode(play) &&
-        (player->meleeWeaponAnimation == 22 || player->meleeWeaponAnimation == 23)) {
+    // #region SOH [Co-op]
+    if (((this->collider.base.acFlags & AC_HIT) && !Player_InCsMode(play) &&
+        (player->meleeWeaponAnimation == 22 || player->meleeWeaponAnimation == 23)) || Flags_GetSwitch(play, this->switchFlag)) {
+    // #endregion
```
Hook material

aMannus commented on soh/src/overlays/actors/ovl_Bg_Hidan_Hamstep/z_bg_hidan_hamstep.c:
```diff
@@ -278,7 +278,9 @@ void func_80888734(BgHidanHamstep* this) {
 
 void func_808887C4(BgHidanHamstep* this, PlayState* play) {
-    if (this->collider.base.acFlags & AC_HIT) {
+    // #region SOH [Co-op]
+    if ((this->collider.base.acFlags & AC_HIT) || Flags_GetSwitch(play, (this->dyna.actor.params >> 8) & 0xFF)) {
+    // #endregion
```
Hook material

aMannus commented on soh/src/overlays/actors/ovl_Bg_Hidan_Hrock/z_bg_hidan_hrock.c:
```diff
@@ -201,7 +201,9 @@ void func_8088960C(BgHidanHrock* this, PlayState* play) {
 
 void func_808896B8(BgHidanHrock* this, PlayState* play) {
-    if (this->collider.base.acFlags & 2) {
+    // #region SOH [Co-op]
+    if ((this->collider.base.acFlags & 2) || Flags_GetSwitch(play, this->unk_16A)) {
+    // #endregion
```
Hook material

aMannus commented on soh/src/overlays/actors/ovl_Bg_Hidan_Kowarerukabe/z_bg_hidan_kowarerukabe.c:
```diff
@@ -303,7 +303,9 @@ void BgHidanKowarerukabe_Update(Actor* thisx, PlayState* play) {
     s32 pad;
 
-    if (Actor_GetCollidedExplosive(play, &this->collider.base) != NULL) {
+    // #region SOH [Co-op]
+    if ((Actor_GetCollidedExplosive(play, &this->collider.base) != NULL) || Flags_GetSwitch(play, (this->dyna.actor.params >> 8) & 0x3F)) {
+    // #endregion
```
Hook material

aMannus commented on soh/src/overlays/actors/ovl_Bg_Ice_Shelter/z_bg_ice_shelter.c:
```diff
@@ -333,10 +333,12 @@ void func_8089107C(BgIceShelter* this, PlayState* play) {
     // Default blue fire check
-    if (this->cylinder1.base.acFlags & AC_HIT) {
+    // #region SOH [Co-op]
+    if ((this->cylinder1.base.acFlags & AC_HIT) || Flags_GetSwitch(play, this->dyna.actor.params & 0x3F)) {
         this->cylinder1.base.acFlags &= ~AC_HIT;
 
-        if ((this->cylinder1.base.ac != NULL) && (this->cylinder1.base.ac->id == ACTOR_EN_ICE_HONO)) {
+        if (((this->cylinder1.base.ac != NULL) && (this->cylinder1.base.ac->id == ACTOR_EN_ICE_HONO)) || Flags_GetSwitch(play, this->dyna.actor.params & 0x3F)) {
+    // #endregion
```
Hook material

aMannus commented on soh/src/overlays/actors/ovl_Bg_Jya_Bombchuiwa/z_bg_jya_bombchuiwa.c:
```diff
@@ -142,7 +142,9 @@ void BgJyaBombchuiwa_SetupWaitForExplosion(BgJyaBombchuiwa* this, PlayState* pla
 
 void BgJyaBombchuiwa_WaitForExplosion(BgJyaBombchuiwa* this, PlayState* play) {
-    if ((this->collider.base.acFlags & AC_HIT) || (this->timer > 0)) {
+    // #region SOH [Co-op]
+    if (((this->collider.base.acFlags & AC_HIT) || (this->timer > 0)) || Flags_GetSwitch(play, this->actor.params & 0x3F)) {
+    // #endregion
```
Hook material

aMannus commented on soh/src/overlays/actors/ovl_Bg_Jya_Bombiwa/z_bg_jya_bombiwa.c:
```diff
@@ -163,7 +163,9 @@ void BgJyaBombiwa_Break(BgJyaBombiwa* this, PlayState* play) {
     BgJyaBombiwa* this = (BgJyaBombiwa*)thisx;
 
-    if (this->collider.base.acFlags & AC_HIT) {
+    // #region SOH [Co-op]
+    if ((this->collider.base.acFlags & AC_HIT) || Flags_GetSwitch(play, this->dyna.actor.params & 0x3F)) {
+    // #endregion
```
Hook material

aMannus commented on soh/src/overlays/actors/ovl_Bg_Mizu_Bwall/z_bg_mizu_bwall.c:
```diff
@@ -467,7 +467,9 @@ void BgMizuBwall_SpawnDebris(BgMizuBwall* this, PlayState* play) {
 void BgMizuBwall_Idle(BgMizuBwall* this, PlayState* play) {
     BgMizuBwall_SetAlpha(this, play);
-    if (this->collider.base.acFlags & AC_HIT) {
+    // #region SOH [Co-op]
+    if ((this->collider.base.acFlags & AC_HIT) || Flags_GetSwitch(play, ((u16)this->dyna.actor.params >> 8) & 0x3F)) {
+    // #endregion
```
Hook material

aMannus commented on soh/src/overlays/actors/ovl_Bg_Spot08_Bakudankabe/z_bg_spot08_bakudankabe.c:
```diff
@@ -183,7 +183,9 @@ void BgSpot08Bakudankabe_Destroy(Actor* thisx, PlayState* play) {
     BgSpot08Bakudankabe* this = (BgSpot08Bakudankabe*)thisx;
 
-    if (this->collider.base.acFlags & AC_HIT) {
+    // #region SOH [Co-op]
+    if ((this->collider.base.acFlags & AC_HIT) || Flags_GetSwitch(play, (this->dyna.actor.params & 0x3F))) {
+    // #endregion
```
Hook material

aMannus commented on soh/src/overlays/actors/ovl_Bg_Spot11_Bakudankabe/z_bg_spot11_bakudankabe.c:
```diff
@@ -135,7 +135,9 @@ void BgSpot11Bakudankabe_Destroy(Actor* thisx, PlayState* play) {
 
-    if (this->collider.base.acFlags & AC_HIT) {
+    // #region SOH [Co-op]
+    if ((this->collider.base.acFlags & AC_HIT) || Flags_GetSwitch(play, (this->dyna.actor.params & 0x3F))) {
+    // #endregion
         func_808B2218(this, play);
```
Hook material

aMannus commented on soh/src/overlays/actors/ovl_Bg_Spot17_Bakudankabe/z_bg_spot17_bakudankabe.c:
```diff
@@ -114,7 +114,9 @@ void BgSpot17Bakudankabe_Destroy(Actor* thisx, PlayState* play) {
 void BgSpot17Bakudankabe_Update(Actor* thisx, PlayState* play) {
     BgSpot17Bakudankabe* this = (BgSpot17Bakudankabe*)thisx;
-    if (this->dyna.actor.xzDistToPlayer < 650.0f && func_80033684(play, &this->dyna.actor) != NULL) {
+    // #region SOH [Co-op]
+    if ((this->dyna.actor.xzDistToPlayer < 650.0f && func_80033684(play, &this->dyna.actor) != NULL) || Flags_GetSwitch(play, (this->dyna.actor.params & 0x3F))) {
+    // #endregion
```
Hook material

aMannus commented on soh/src/overlays/actors/ovl_Bg_Ydan_Maruta/z_bg_ydan_maruta.c:
```diff
@@ -146,7 +146,9 @@ void func_808BEFF4(BgYdanMaruta* this, PlayState* play) {
 
 void func_808BF078(BgYdanMaruta* this, PlayState* play) {
-    if (this->collider.base.acFlags & AC_HIT) {
+    // #region SOH [Co-op]
+    if ((this->collider.base.acFlags & AC_HIT) || Flags_GetSwitch(play, this->switchFlag)) {
+    // #endregion
```
Hook material

aMannus commented on soh/src/overlays/actors/ovl_Bg_Ydan_Sp/z_bg_ydan_sp.c:
```diff
@@ -282,6 +282,12 @@ void BgYdanSp_FloorWebIdle(BgYdanSp* this, PlayState* play) {
     webPos.x = this->dyna.actor.world.pos.x;
     webPos.y = this->dyna.actor.world.pos.y - 50.0f;
     webPos.z = this->dyna.actor.world.pos.z;
+    // #region SOH [Co-op]
+    if (Flags_GetSwitch(play, this->isDestroyedSwitchFlag)) {
+        BgYdanSp_BurnWeb(this, play);
+        return;
+    }
+    // #endregion
```
Hook material

aMannus commented on soh/src/overlays/actors/ovl_Bg_Ydan_Sp/z_bg_ydan_sp.c:
```diff
@@ -411,6 +417,11 @@ void BgYdanSp_WallWebIdle(BgYdanSp* this, PlayState* play) {
             BgYdanSp_BurnWeb(this, play);
         }
     }
+    // #region SOH [Co-op]
+    if (Flags_GetSwitch(play, this->isDestroyedSwitchFlag)) {
+        BgYdanSp_BurnWeb(this, play);
+    }
+    // #endregion
```
Hook material

aMannus commented on soh/src/overlays/actors/ovl_Door_Shutter/z_door_shutter.c:
```diff
@@ -645,6 +646,11 @@ void DoorShutter_Update(Actor* thisx, PlayState* play) {
     if (!(player->stateFlags1 & (PLAYER_STATE1_TALKING | PLAYER_STATE1_DEAD | PLAYER_STATE1_GETTING_ITEM | PLAYER_STATE1_IN_ITEM_CS)) || (this->actionFunc == DoorShutter_SetupType)) {
         this->actionFunc(this, play);
     }
+    // #region SOH [Co-op]
+    if (Flags_GetSwitch(play, this->dyna.actor.params & 0x3F)) {
+        DECR(this->unk_16E);
+    }
+    // #endregion
```
Hook material

aMannus commented on soh/src/overlays/actors/ovl_En_Door/z_en_door.c:
```diff
@@ -234,6 +235,11 @@ void EnDoor_Idle(EnDoor* this, PlayState* play) {
             this->actionFunc = EnDoor_AjarOpen;
         }
     }
+    // #region SOH [Co-op]
+    if (Flags_GetSwitch(play, this->actor.params & 0x3F)) {
+        DECR(this->lockTimer);
+    }
+    // #endregion
```
Hook material

aMannus commented on soh/src/overlays/actors/ovl_En_Si/z_en_si.c:
```diff
@@ -149,6 +149,13 @@ void func_80AFB950(EnSi* this, PlayState* play) {
 void EnSi_Update(Actor* thisx, PlayState* play) {
     EnSi* this = (EnSi*)thisx;
 
+    // #region SOH [Co-op]
+    if (GET_GS_FLAGS((thisx->params & 0x1F00) >> 8) & (thisx->params & 0xFF)) {
+        Actor_Kill(&this->actor);
+        return;
+    }
+    // #endregion
+
```
Hook material

aMannus commented on soh/src/overlays/actors/ovl_En_Sw/z_en_sw.c:
```diff
@@ -898,6 +898,13 @@ void func_80B0E9BC(EnSw* this, PlayState* play) {
 void EnSw_Update(Actor* thisx, PlayState* play) {
     EnSw* this = (EnSw*)thisx;
 
+    // #region SOH [Co-op]
+    if (GET_GS_FLAGS((thisx->params & 0x1F00) >> 8) & (thisx->params & 0xFF)) {
+        Actor_Kill(&this->actor);
+        return;
+    }
+    // #endregion
+
```
Hook material

aMannus commented on soh/src/overlays/actors/ovl_Item_B_Heart/z_item_b_heart.c:
```diff
@@ -55,6 +55,13 @@ void ItemBHeart_Destroy(Actor* thisx, PlayState* play) {
 void ItemBHeart_Update(Actor* thisx, PlayState* play) {
     ItemBHeart* this = (ItemBHeart*)thisx;
 
+    // #region SOH [Co-op]
+    if (Flags_GetCollectible(play, 0x1F)) {
+        Actor_Kill(&this->actor);
+        return;
+    }
+    // #endregion
+
```
Hook material

aMannus commented on soh/src/overlays/actors/ovl_Obj_Bombiwa/z_obj_bombiwa.c:
```diff
@@ -125,8 +125,10 @@ void ObjBombiwa_Update(Actor* thisx, PlayState* play) {
     s32 pad;
 
+    // #region SOH [Co-op]
     if ((func_80033684(play, &this->actor) != NULL) ||
-        ((this->collider.base.acFlags & AC_HIT) && (this->collider.info.acHitInfo->toucher.dmgFlags & 0x40000040))) {
+        ((this->collider.base.acFlags & AC_HIT) && (this->collider.info.acHitInfo->toucher.dmgFlags & 0x40000040)) || Flags_GetSwitch(play, this->actor.params & 0x3F)) {
+    // #endregion
```
Hook material

aMannus commented on soh/src/overlays/actors/ovl_Obj_Hamishi/z_obj_hamishi.c:
```diff
@@ -171,10 +171,12 @@ void ObjHamishi_Update(Actor* thisx, PlayState* play) {
 
-    if ((this->collider.base.acFlags & AC_HIT) && (this->collider.info.acHitInfo->toucher.dmgFlags & 0x40000040)) {
+    // #region SOH [Co-op]
+    if (((this->collider.base.acFlags & AC_HIT) && (this->collider.info.acHitInfo->toucher.dmgFlags & 0x40000040)) || Flags_GetSwitch(play, this->actor.params & 0x3F)) {
         this->collider.base.acFlags &= ~AC_HIT;
         this->hitCount++;
-        if (this->hitCount < 2) {
+        if (this->hitCount < 2 && !Flags_GetSwitch(play, this->actor.params & 0x3F)) {
+    // #endregion
```
Hook material

aMannus reviewed (commented):
Want to bring up the naming point, and also tagged all the vanilla code changes. Most of these will likely just become a VB_Should. Might help with the effort to bring them over later, just figured it'd be nice to have an indication of how many there are and a list where we can cross them off as they are changed.

Fatih120 commented:
Hello, why is it that when me and my friend are trying to play a Randomizer Seed together, the person receiving the others' JSON is unable to play? The save selection entry is greyed out after creating that save, and thus we can't start and join each other.
![image](https://github.com/user-attachments/assets/a96fc1ff-5896-4243-814a-d1100053d3fa)

Malkierian commented:
This is vanila/MQ mismatch. One of you doesn't have the same game versions the other does, and if a seed is generated with dungeon types a person doesn't have access to because they haven't extracted that version of the game (i.e. you have both vanilla and MQ and mix dungeon types, and send to the other person who only has vanilla), then you can't play the seed.

garrettjoecox commented:
This PR has been updated with various fixes, and all vanilla code touch points have been migrated to hooks.

The known remaining issues are listed at the bottom of the PR.

garrettjoecox committed:
Anchor

This is far and away the proudest thing I have been apart of in the Zelda community. Thank you to everyone who either directly or indirectly contributed to this effort. If I forgot to co-author you and you were involved, know that I simply forgot but still appreciated your help! Also thank you to the literal tens of thousands of people who have played Anchor, providing feedback and bug reports. Super thrilled to finally have this merged into the main Ship of Harkinian experience.

Co-authored-by: mattman107 <65982675+mattman107@users.noreply.github.com>
Co-authored-by: David Chavez <david@dcvz.io>
Co-authored-by: MelonSpeedruns <melonspeedruns@outlook.com>
Co-authored-by: aMannus <mannusmenting@gmail.com>
Co-authored-by: Malkierian <malkierian@gmail.com>
Co-authored-by: Caladius <Caladius@users.noreply.github.com>
Co-authored-by: Patrick12115 <115201185+Patrick12115@users.noreply.github.com>
Co-authored-by: PurpleHato <47987542+PurpleHato@users.noreply.github.com>
Co-authored-by: balloondude2 <55861555+balloondude2@users.noreply.github.com>
Co-authored-by: lilacLunatic <8488221+lilacLunatic@users.noreply.github.com>
Co-authored-by: Felix Lee <flee135@users.noreply.github.com>
Co-authored-by: Sirius902 <10891979+Sirius902@users.noreply.github.com>

Malkierian reviewed (approved):
Honestly, with the footprint into existing code being so miniscule now, and with the playtesting that has been ongoing for so long, I think this can just be merged. I'll give anyone who disagrees 2 days to voice it.
